DSA/1.1 Arrays/3. Range Queries/1.1.3.5 Arrays: Range Queries.md

```diff
@@ -0,0 +1,40 @@
+<h1 align="center">Suffix Array</h1>
+<ul>
+  <li>A suffix array is a sorted array of all suffixes of a given string.Any suffix tree based algorithm can be replaced with an algorithm that uses a suffix array enhanced with additional information and solves the same problem in the same time complexity</p></li>
+  <li>A suffix array can be constructed from Suffix tree by doing a DFS traversal of the suffix tree. In fact Suffix array and suffix tree both can be constructed from each other in linear time.</li>
+  <li>Advantages of suffix arrays over suffix trees include improved space requirements, simpler linear time construction algorithms and improved cache locality</li>
+</ul>  
+<h2>Suffix Tree</h2>
+<p>A <b>Suffix tree</b> (also called PAT tree or, in an earlier form, position tree) is a compressed trie containing all the suffixes of the given text as their keys and positions in the text as their values. Suffix trees allow particularly fast implementations of many important string operations.</p>
+<h3>The suffix tree for the string S of length n is defined as a tree such that:</h3>
+<ul>
+  <li>The tree has exactly n leaves numbered from 1 to n.</li>
+  <li>Except for the root, every internal node has at least two children.</li>
+  <li>Each edge is labelled with a non-empty substring of S.</li>
+  <li>No two edges starting out of a node can have string-labels beginning with the same character.</li>
+  <li>he string obtained by concatenating all the string-labels found on the path from the root to leaf i spells out suffix S[i..n], for i from 1 to n.</li>
+</ul>  
+<h3>Example:</h3>
+<pre>{bear, bell, bid, bull, buy, sell, stock, stop}</pre>
+<p>Following is standard trie for the above input set of words.</p>
+<div align="center">
+    <img width="45%" alt="StandardTrie" src="https://media.geeksforgeeks.org/wp-content/uploads/patternsearching1.png">
+</div>
+<p>Following is the compressed trie. Compress Trie is obtained from standard trie by joining chains of single nodes. The nodes of a compressed trie can be stored by storing index ranges at the nodes.</p>
+<div align="center">
+    <img width="45%" alt="CompressedTrie" src="https://media.geeksforgeeks.org/wp-content/uploads/patternsearching2.png">
+</div>
+<h3>How to build a Suffix Tree for a given text?</h3>
+<p>As discussed above, Suffix Tree is compressed trie of all suffixes, so following are very abstract steps to build a suffix tree from given text.</p>
+<ol type="1">
+  <li>Generate all suffixes of given text.</li>
+  <li>Consider all suffixes as individual words and build a compressed trie.</li>
```

Comment:
mention something about algorithm by Ukkonen for building suffix tree

Reply:
Ok I will add it too